api/src/main/java/bio/terra/tanagra/serialization/UFDisplayHint.java

```diff
@@ -0,0 +1,55 @@
+package bio.terra.tanagra.serialization;
+
+import bio.terra.tanagra.serialization.displayhint.UFEnumVals;
+import bio.terra.tanagra.serialization.displayhint.UFNumericRange;
+import bio.terra.tanagra.underlay.DisplayHint;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
+
+/**
+ * External representation of an attribute display hint.
+ *
+ * <p>This is a POJO class intended for serialization. This JSON format is user-facing.
+ */
+@JsonTypeInfo(
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.EXISTING_PROPERTY,
+    property = "type")
+@JsonSubTypes({
+  @JsonSubTypes.Type(value = UFEnumVals.class, name = "ENUM"),
+  @JsonSubTypes.Type(value = UFNumericRange.class, name = "RANGE")
+})
+@JsonDeserialize(builder = UFDisplayHint.Builder.class)
+public abstract class UFDisplayHint {
+  private final DisplayHint.Type type;
+
+  protected UFDisplayHint(DisplayHint displayHint) {
+    this.type = displayHint.getType();
+  }
+
+  protected UFDisplayHint(Builder builder) {
+    this.type = builder.type;
+  }
+
+  @JsonPOJOBuilder(buildMethodName = "build", withPrefix = "")
+  public abstract static class Builder {
+    private DisplayHint.Type type;
+
+    public Builder type(DisplayHint.Type type) {
+      this.type = type;
+      return this;
+    }
+
+    /** Call the private constructor. */
+    public abstract UFDisplayHint build();
+  }
+
+  /** Deserialize to the internal representation of the display hint. */
+  public abstract DisplayHint deserializeToInternal();
```

Comment:
Yes this is a little hard to follow, but I'm not sure how to improve it. I used this same approach for the data pointer classes (see `UF/DataPointer` and `UF/BigQueryDataset`), so the below applies also to that case and any future uses of inheritance.

Both the `serialization` package and the internal `underlay` package contain a base class `UF/DisplayHint` and sub-classes e.g. `UF/NumericRange`. I need to be able to convert in both directions, and I would like to avoid using a `switch(type)` block, since that sort of defeats the purpose of using inheritance. So I used an abstract method on both classes: `serialize` on the internal class, and `deserializeToInternal` on the serialization class.

The deserialization logic includes expanding defaults and pulling information from higher-up objects (e.g. table name, BQ dataset id), so it's significantly more complicated than the serialization logic. I agree with you that it makes more sense to keep the serialized versions as simple JSON mappings and put any non-trivial logic into the internal classes. That's why all the internal objects have a `fromSerialized` method. For sets of classes that use inheritance, there's the additional step of calling the `fromSerialized` method from the serialization class for the reason in the first paragraph.